modules/apps/client-extension/client-extension-api/src/main/java/com/liferay/client/extension/util/CETUtil.java

```diff
@@ -5,14 +5,35 @@
 
 package com.liferay.client.extension.util;
 
+import com.liferay.client.extension.constants.ClientExtensionEntryConstants;
+import com.liferay.client.extension.model.ClientExtensionEntryRel;
+import com.liferay.client.extension.service.ClientExtensionEntryRelLocalServiceUtil;
 import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.model.LayoutSet;
+import com.liferay.portal.kernel.util.PortalUtil;
 import com.liferay.portal.kernel.util.Validator;
 
 /**
  * @author Binh Tran
  */
 public class CETUtil {
 
+	public static String getLayoutSetThemeCSSCETExternalReferenceCode(
```

Comment:
@ethib137 and I were discussing regarding where to put this method. It's called during page load to fetch the `FrontendTokenDefinition` (see `FrontendTokenDefinitionRegistryImpl`). The concern is having a type specific method (`ThemeCSSCET`) on a generic util `CETUtil`.

What do you guys think? @dsanz @izaera

Reply:
I'd say we can leverage [ClientExtensionServicePreAction](https://github.com/liferay/liferay-portal/blob/6ce0b9eb049e99a50285a324a3bd4afd76d340ab/modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/internal/events/ClientExtensionsServicePreAction.java) for this purpose. Indeed, it's already fetching the right ThemeCSSCET object, considering the fact that one can set a ThemeCSSCET at page level too, and not just at layoutset level.

Perhaps, that would be the point where the right ThemeCSSCET object is set in a place where it can be found later on. Possible places would be the request or the theme display. We'd need to see where this info is needed, to make sure any of the above is available.

Reply:
If we're still going to have a scenario where the ThemeCSSCET wouldn't be available, couldn't we just change the interface of the `FrontendTokenDefinitionRegistry#getFrontendTokenDefinition`? Currently is 

```
public FrontendTokenDefinition getFrontendTokenDefinition(
		long companyId, String externalReferenceCode, String themeId);
```
would now be

```
public FrontendTokenDefinition getFrontendTokenDefinition(Layout layout) {
    LayoutSet layoutSet = layout.getLayoutSet();
    // gets the external reference code based on that
}
```

but once we handle ThemeCSSCET for page level (refer to [Evan's comment](https://github.com/liferay-platform-experience/liferay-portal/pull/467#discussion_r1481133337)), we do something like this

```
public FrontendTokenDefinition getFrontendTokenDefinition(Layout layout) {
        CET cet = _getCET(
		_portal.getClassNameId(Layout.class), layout.getPlid(),
		layout.getCompanyId(),
		ClientExtensionEntryConstants.TYPE_THEME_CSS);

	if (cet == null) {
		cet = _getCET(
			_portal.getClassNameId(Layout.class),
			layout.getMasterLayoutPlid(), layout.getCompanyId(),
			ClientExtensionEntryConstants.TYPE_THEME_CSS);
	}

	if (cet == null) {
		LayoutSet layoutSet = layout.getLayoutSet();

		cet = _getCET(
			_portal.getClassNameId(LayoutSet.class),
			layoutSet.getLayoutSetId(), layout.getCompanyId(),
			ClientExtensionEntryConstants.TYPE_THEME_CSS);
	}
	// ...
}
```

Reply:
Is it reasonable or am I missing something?

Reply:
I think this is reasonable, as you're adapting the FTDRegistry to be able to return FTD in a variety of scenarios